src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines
+{
+    internal class SequencePipeReader : PipeReader
+    {
+        private ReadOnlySequence<byte> _sequence;
+        private bool _isReaderCompleted;
+
+        private int _cancelNext;
+
+        public SequencePipeReader(ReadOnlySequence<byte> sequence)
+        {
+            _sequence = sequence;
+        }
+
+        /// <inheritdoc />
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            AdvanceTo(consumed, consumed);
+        }
+
+        /// <inheritdoc />
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            ThrowIfCompleted();
+
+            // Fast path: did we consume everything?
+            if (consumed.Equals(_sequence.End))
+            {
+                _sequence = ReadOnlySequence<byte>.Empty;
+                return;
+            }
+
+            _sequence = _sequence.Slice(consumed);
+        }
+
+        /// <inheritdoc />
+        public override void CancelPendingRead()
+        {
+            Interlocked.Exchange(ref _cancelNext, 1);
+        }
+
+        /// <inheritdoc />
+        public override void Complete(Exception? exception = null)
+        {
+            if (_isReaderCompleted)
+            {
+                return;
+            }
+
+            _isReaderCompleted = true;
+            _sequence = ReadOnlySequence<byte>.Empty;
+        }
+
+        /// <inheritdoc />
+        public override ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            ThrowIfCompleted();
+
+            if (TryReadInternal(out var result))
+            {
+                return new ValueTask<ReadResult>(result);
+            }
+
+            result = new ReadResult(ReadOnlySequence<byte>.Empty, isCanceled: false, isCompleted: true);
+            return new ValueTask<ReadResult>(result);
+        }
+
+        /// <inheritdoc />
+        public override bool TryRead(out ReadResult result)
+        {
+            ThrowIfCompleted();
+
+            return TryReadInternal(out result);
+        }
+
+        private bool TryReadInternal(out ReadResult result)
```

Comment:
Why is this method needed?  Why not just have all of this in TryRead and then have ReadAsync call TryRead?

Reply:
2 reasons. Firstly there used to be a bigger difference if I recall correctly (which isn't a reason to keep it), and second it mirrors the implementation for a stream pipe reader. That being said, as it is now I agree that it should probably just be removed, and I will do so.